### PR TITLE
[MAINT] Fixed Handling Negative Values for P2S

### PR DESCRIPTION
**Pull request by ShreyasFadnavis**
Small fix as an alternative to clipping to a minimum via `np.clip` !

### CODE REVIEW

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @ShreyasFadnavis,

Thank you for this fix. I think you need to review your change or update your test. This change make your test less robust. See below the error:

```python
    @needs_sklearn
    def test_patch2self_random_noise():
        S0 = 30 + 2 * np.random.standard_normal((20, 20, 20, 50))
    
        bvals = np.repeat(30, 50)
        S0nb = p2s.patch2self(S0, bvals, model='ols')
    
>       assert_greater(S0nb.min(), S0.min())
E           AssertionError: 21.1010317819 > 21.1010317819
```

**Comment by ShreyasFadnavis** [comment]
> Hi @ShreyasFadnavis,
> 
> Thank you for this fix. I think you need to review your change or update your test. This change make your test less robust. See below the error:
> 
> ```python
>     @needs_sklearn
>     def test_patch2self_random_noise():
>         S0 = 30 + 2 * np.random.standard_normal((20, 20, 20, 50))
>     
>         bvals = np.repeat(30, 50)
>         S0nb = p2s.patch2self(S0, bvals, model='ols')
>     
> >       assert_greater(S0nb.min(), S0.min())
> 
> E           AssertionError: 21.1010317819 > 21.1010317819
> ```

Thanks @skoudoro ... On it!

**Review by skoudoro — COMMENTED** [review]
It seems ready to go. It will be good to have another opinion before merging

**Comment by arokem** [comment]
Generally looks good. Just want to make sure that I understand what this is for: you are trying to preserve the variance, while avoiding negative values?

**Comment by ShreyasFadnavis** [comment]
> Generally looks good. Just want to make sure that I understand what this is for: you are trying to preserve the variance, while avoiding negative values?

Yes exactly! This was pointed out to me by @RafaelNH one of his datasets, where this issue was messing up the `b0`'s. I will add this as an option than forcing it always.

**Comment by ShreyasFadnavis** [comment]
Okay! I am done 👍🏽 Let me know if something seems off!

File: dipy/denoise/patch2self.py [hunk 1]
```diff
@@ -315,8 +317,18 @@ def patch2self(data, bvals, patch_radius=[0, 0, 0], model='ridge',
         denoised_arr[:, :, :, idx[0]] = np.squeeze(denoised_dwi[..., i])
 
     # shift intensities per volume to handle for negative intensities
-    for i in range(0, denoised_arr.shape[3]):
-        shift = np.min(data[..., i]) - np.min(denoised_arr[..., i])
-        denoised_arr[..., i] = denoised_arr[..., i] + shift
+    if shift_intensity == True and clip_negative_vals == False:
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py]: in python, for a boolean value, you should avoid `==True` and `is True`. Can you update that everywhere?

In this specific case, `if shift_intensity and not clip_negative_vals:`

For your information, a quick test:
```python
In [1]: 1 is True                                                                                                                                                                                            
Out[1]: False

In [2]: 1 == True                                                                                                                                                                                            
Out[2]: True
```

File: dipy/denoise/patch2self.py [hunk 2]
```diff
@@ -256,9 +258,9 @@ def patch2self(data, bvals, patch_radius=[0, 0, 0], model='ridge',
         t1 = time.time()
 
     # if only 1 b0 volume, skip denoising it
-    if data_b0s.ndim == 3:
+    if data_b0s.ndim == 3 or b0_denoising is False:
         if verbose is True:
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py:276]: `if verbose` instead

File: dipy/denoise/patch2self.py [hunk 3]
```diff
@@ -315,8 +317,18 @@ def patch2self(data, bvals, patch_radius=[0, 0, 0], model='ridge',
...
-    for i in range(0, denoised_arr.shape[3]):
-        shift = np.min(data[..., i]) - np.min(denoised_arr[..., i])
-        denoised_arr[..., i] = denoised_arr[..., i] + shift
+    if shift_intensity == True and clip_negative_vals == False:
+        for i in range(0, denoised_arr.shape[3]):
+            shift = np.min(data[..., i]) - np.min(denoised_arr[..., i])
+            denoised_arr[..., i] = denoised_arr[..., i] + shift
+
+    # clip out the negative values from the denoised output
+    elif clip_negative_vals == True and shift_intensity == False:
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py]: same comment as above

File: dipy/denoise/patch2self.py [hunk 4]
```diff
@@ -315,8 +317,18 @@ def patch2self(data, bvals, patch_radius=[0, 0, 0], model='ridge',
...
+        denoised_arr.clip(min=0, out=denoised_arr)
+
+    elif clip_negative_vals == True and shift_intensity == True:
```
Comment (skoudoro) [inline dipy/denoise/patch2self.py]: same comment as above

**Review by skoudoro — COMMENTED** [review]
Hi @ShreyasFadnavis,

See below for some comments. is it possible to add a test for each case to make sure the expected behavior appears. Thank you

File: dipy/denoise/tests/test_patch2self.py [hunk 5]
```diff
@@ -17,11 +17,16 @@ def test_patch2self_random_noise():
     S0 = 30 + 2 * np.random.standard_normal((20, 20, 20, 50))
 
     bvals = np.repeat(30, 50)
-    S0nb = p2s.patch2self(S0, bvals, model='ols')
+    S0den_shift = p2s.patch2self(S0, bvals, model='ols', shift_intensity=True)
```
Comment (ShreyasFadnavis) [inline dipy/denoise/tests/test_patch2self.py:20]: @skoudoro the test should be handled here 👍🏽

**Review by skoudoro — APPROVED** [review]
Thank you for the update @ShreyasFadnavis.

Looks good to me, waiting for the CI's and any other feedback?

**Comment by Garyfallidis** [comment]
+1 from me just restarted some bots. If these are okay then please merge.

**Pull request merged**